htlcswitch/link.go

```diff
@@ -638,24 +748,66 @@ func (l *channelLink) RemoveFlushHook(FlushHookID) error {
 // LinkDirection, then we will call this hook synchronously and will not return
 // a CommitHookID.
 func (l *channelLink) OnCommitOnce(
-	LinkDirection, func(),
+	direction LinkDirection, hook func(),
 ) fn.Option[CommitHookID] {
 
-	// TODO(proofofkeags): Implement
-	return fn.None[CommitHookID]()
+	// If we want to run the hook once on the next fully committed state,
+	// and we don't currently need/owe a commitment, it means we don't have
+	// any pending updates and we can run the hook synchronously. If we call
+	// it syncrhonously we will not add it to the map at all.
+	runSync := direction == Outgoing && !l.channel.OweCommitment() ||
+		direction == Incoming && !l.channel.NeedCommitment()
+	if runSync {
+		hook()
```

Comment:
I don't think this can be run synchronously because the callback is not invoked from the `htlcManager` goroutine.  If the link sends an htlc after the value of `runSync` is set but before `hook()` is run, we'll end up violating the protocol

Reply:
What if we `RLock` the link here?

Reply:
We currently Lock link state for this so I think we're OK.

Reply:
We need it though. If we are already in a clean commitment state, and we remove this, it will result in having to wait for our *next* commitment to get sent out. This can result in protocol violations. If you find that the mutex won't suffice we can change this to a different sync model wherein we accept the hooks over channels and then they either get immediately executed in the htlcManager thread or they get added to the map for deferred execution.